publication-commons/src/main/java/no/unit/nva/doi/models/ImmutableDoi.java

```diff
@@ -0,0 +1,396 @@
+
+    @SuppressWarnings("PMD.CallSuperInConstructor")
+    private ImmutableDoi(ImmutableDoi.Builder builder) {
+        this.prefix = builder.prefix;
+        this.suffix = builder.suffix;
+        this.proxy = builder.proxyIsSet()
+            ? builder.proxy
+            : Objects.requireNonNull(super.getProxy(), MESSAGE_NON_NULL_ARGUMENT_FOR_PARAMETER_PROXY);
+    }
+
+    @SuppressWarnings("PMD.CallSuperInConstructor")
+    private ImmutableDoi(URI proxy, String prefix, String suffix) {
+        this.proxy = proxy;
+        this.prefix = prefix;
+        this.suffix = suffix;
+    }
+
+    /**
+     * Creates an immutable copy of a {@link Doi} value. Uses accessors to get values to initialize the new immutable
+     * instance. If an instance is already immutable, it is returned as is.
+     *
+     * @param instance The instance to copy
+     * @return A copied immutable Doi instance
+     */
+    public static ImmutableDoi copyOf(Doi instance) {
+        if (instance instanceof ImmutableDoi) {
+            return (ImmutableDoi) instance;
+        }
+        return ImmutableDoi.builder()
+            .withProxy(instance.getProxy())
+            .withPrefix(instance.getPrefix())
+            .withSuffix(instance.getSuffix())
+            .build();
+    }
+
+    /**
+     * Creates a builder for {@link ImmutableDoi ImmutableDoi}.
+     * <pre>
+     * ImmutableDoi.builder()
+     *    .proxy(String) // optional {@link Doi#getProxy() proxy}
+     *    .prefix(String) // required {@link Doi#getPrefix()} () prefix}
+     *    .suffix(String) // required {@link Doi#getSuffix()} () suffix}
+     *    .build();
+     * </pre>
+     *
+     * @return A new ImmutableDoi builder
+     */
+    public static ImmutableDoi.Builder builder() {
+        return new ImmutableDoi.Builder();
+    }
+
+    /**
+     * Retrieve the value of the ${@code proxy} attribute.
+     *
+     * @return The value of the {@code proxy} attribute
+     */
+    @Override
+    public URI getProxy() {
+        return proxy;
+    }
+
+    /**
+     * Retrieve the value of the ${@code prefix} attribute.
+     *
+     * @return The value of the {@code prefix} attribute
+     */
+    @Override
+    public String getPrefix() {
+        return prefix;
+    }
+
+    /**
+     * Retrieve the value of the ${@code suffix} attribute.
+     *
+     * @return The value of the {@code suffix} attribute
+     */
+    @Override
+    public String getSuffix() {
+        return suffix;
+    }
+
+    /**
+     * Copy the current immutable object by setting a value for the {@link Doi#getProxy() proxy} attribute. An equals
+     * check used to prevent copying of the same value by returning {@code this}.
+     *
+     * @param value A new value for proxy
+     * @return A modified copy of the {@code this} object
+     */
+    public final ImmutableDoi withProxy(URI value) {
+        URI newValue = Objects.requireNonNull(value, MESSAGE_NON_NULL_ARGUMENT_FOR_PARAMETER_PROXY);
```

Comment:
These methods are problematic because they replicate the logic in the builder. It could be much simpler if the builder allowed overwriting values, copy would work by creating a builder and validation happened in build. 
```
Doi existingDoi;
doi.copy().withSuffix(newSuffix).build().
```
Also for the time being I do not see the need to have two classes for Doi. One class where one creates a Doi and cannot mutate the fields (i.e., no setters) would be OK.

Reply:
Removed `copyOf` functionality